---
layout: exercise
topic: Genomics
title: Sequence Quality Trimming
language: Shell
---
Uncomment below to work on exercise

<!--

> ## Exercise 1
>
> Use the output from your Trimmomatic command to answer the
> following questions.
>
> 1) What percent of reads did we discard from our sample?
> 2) What percent of reads did we keep both pairs?
>

> ## Exercise 2
> We trimmed our fastq files with Nextera adapters, 
> but there are other adapters that are commonly used.
> What other adapter files came with Trimmomatic?
>


> ## Bonus Exercise (Advanced)
>
> Now that we've quality controlled our samples, they should perform
> better on the quality tests run by FastQC. Go ahead and re-run
> FastQC on your trimmed FASTQ files and visualize the HTML files
> to see whether your per base sequence quality is higher after
> trimming.

-->

Genomics-trimming-and-filtering-Shell.md